scalafix/rules/src/main/scala/fix/NoWithForExtends.scala

```diff
@@ -0,0 +1,24 @@
+package fix
+
+import scalafix.v1._
+
+import scala.meta._
+import scala.meta.tokens.Token.{KwWith, KwExtends}
+
+case class WithForExtends(pos: Position) extends Diagnostic {
+  override def message = "The `with` keyword is unnecessary here, replace with a comma"
+  override def position: _root_.scala.meta.Position = pos
+}
+
+class NoWithForExtends extends SyntacticRule("NoWithForExtends") {
+  @annotation.tailrec
+  private def containsSublist(list: List[Token]): Option[Token] = list match {
+    case Nil => None
+    case (_: KwExtends) :: _ :: _ :: _ :: (wth: KwWith) :: _ => Some(wth)
```

Comment:
The fact that `Foo extends Bar with Baz` has the same AST as `Foo extends Baz, Baz` is sad 😢 I think matching on `Token`s like this could work with some more logic but unfortunately there are at least a few cases this won't catch:

1. When there are multiple cases of `extends ... with ...` in the same file
2. When the type being extended has type or value parameters, e.g.
    ```scala
    class Foo extends Bar(...) with Baz
    class Foo extends Bar[...] with Baz
    ```
3. When there are multiple `with`s, e.g.
    ```scala
    class Foo extends Bar with Baz with Quux
    ```

What we might be able to do as an alternative is use `doc.tree.collect` to find any `Template`s where `Template#inits > 1`, then inspect each `Template`'s tokens and apply patches where necessary. 

We'll need to be careful to make sure the patches are only applied to the relevant `with` keywords -- maybe we could keep track of whether we're currently inside parentheses (for value params) or square brackets (for type params)? For example:

```scala
trait HasId { val id: Int }
trait HasEmail { val email: String }

trait Foo[A]
trait Baz
class Bar extends Foo[HasId with HasEmail] with Baz
                         // ^^^^ we don't want to lint this `with`
```

And then, assuming we're not in parentheses or square brackets, we could stop looking for more `with`s once we encounter `{` or the last token, whichever comes first -- the former indicating that the body of the class is starting and the latter indicating that there's no body.

Does that seem reasonable? This is definitely more complex than I anticipated so we can also reconsider whether it's worthwhile

Reply:
Hey thanks that was really helpful. I've added some more logic that I think addresses these concerns. One thing I'm not sure on is your concern regarding something like 

`trait Foo extends Bar with Baz with Bat` 

As far as I can see, as long as we match the first `with` we should be good, as the compiler doesn't seem to tolerate mixing `with` and `,` usages here. For example: 

`trait Foo extends Bar, Baz with Bat`

Doesn't compile.